### ISSUE TITLE
create our own parser and node structure

### ISSUE THREAD

**Issue by evykassirer** [socraticorg/mathsteps#128]
using @kevinbarabash's parsing tool https://github.com/kevinbarabash/math-parser

----

### some context from #89:

the proposal:

> I'd like to make a suggestion: replace `mathjs`.  The main parts of `mathjs` that I see we're using are the parser and the eval code.  Both of these are relatively easy to replace considering most of the heavy lifting for evaluation would be done by `fraction.js` anyways.
> 
> Benefits:
> - more easily change which numbers we're computing with, e.g. fractions, complex numbers
> - have a more regular AST, right now we have EquationNode but we're bound to add more as we go
> - avoid re-parsing things over and over
> - have a parser that produces a flattened tree by design
> - smaller bundle size

things that need to be added to `math-parser` to support `mathsteps`:

- [ ] more tests
- [ ] eval
- [ ] anything else?

discussion on `math-parser`: https://github.com/kevinbarabash/math-parser/issues/12

things that need to happen in `mathsteps` to swap out mathjs' expression tree for math-parser's AST:

- [ ] write a function to convert the math-parser AST to match mathjs' expression tree (depends on https://github.com/kevinbarabash/math-parser/issues/8)
- [ ] verify that both match for all expressions that we care about, maybe write a function that traverse either nodes from either tree and produces an s-expression string which can be easily compared
- [ ] update functions throughout `mathsteps` to handle either type of node
- [ ] replace method calls to .eval(), with an evaluate function that can accept either node
- [ ] swap out the parser

**Comment by kevinbarabash** [comment]
@evykassirer thanks for opening this issue.

I've been a little distracted working [math-ast](https://github.com/kevinbarabash/math-ast) which provides a [spec](https://github.com/kevinbarabash/math-ast/blob/master/spec.md) for what a shared AST might look like.  My hope that by defining a more comprehensive a reusable AST that targets K-12 + undergrad math + science we promote the development of interoperable tools from a community of developers.  I'd really appreciate y'all having a look at the spec and opening issues for things that you think are missing or could be improved.  I'll be sharing it more widely eventually.

I'd like [math-parser](https://github.com/kevinbarabash/math-parser) to be the reference implementation of a parser that produces a math-ast structure.  I should actually stop calling it an AST b/c the math-ast spec represents the semantic structure.  I guess AST could stand for Abstract Semantic Tree.

I've created a [transform](https://github.com/kevinbarabash/math-parser/blob/master/lib/mathjs-transform.js) in the math-parser repo that produces a structure similar to what mathjs does.

I'm going to spend some time updating math-parser to produce what's currently in the math-ast spec and then update the transform to produce what it's producing now.  This way we can continue to iterate on the AST spec while maintaining compatibility with mathsteps.

I think the next step for mathsteps would be to create an `evaluate` function that's passed nodes instead of calling `eval` on the node.  Then we can try swapping out the parser and calling math-parser's [evaluate](https://github.com/kevinbarabash/math-parser/blob/master/lib/evaluate.js) function on the nodes.  Of course before any swapping out we need some tests for math-parser.  I'll try to get to those this weekend.

**Comment by kevinbarabash** [comment]
Quick update: I've put up an interactive [demo](https://kevinbarabash.github.io/math-parser/) for math-parser.  You can switch the output format by using the drop down.  Probably the biggest pending issue is https://github.com/kevinbarabash/math-parser/issues/16.  If you discover other issues please use the "permalink" button in the demo and copy a link into into the issue.

**Comment by evykassirer** [comment]
cooool!

Maybe change 'mathsteps' to 'mathjs', since we can't really take any credit for that parsing

I think the easiest way to see if they can be equivalent is if you can map your math-ast to the equivalent mathjs tree, and then put it through mathsteps and see if it gets the same steps as if you parse directly with mathjs

**Comment by kevinbarabash** [comment]
@evykassirer I've made the requested label change in the demo.  I've also added `evaluateMathJS` which evaluates ASTs produced by `transformMathJS` along with [tests](https://github.com/kevinbarabash/math-parser/blob/master/test/mathjs-evaluate_test.js).  https://github.com/kevinbarabash/math-parser/issues/16 has been fixed too.

**Comment by evykassirer** [comment]
okay swapped out eval in #135 - is that what you were thinking?

**Comment by kevinbarabash** [comment]
@evykassirer that's exactly what I was thinking. ^_^

**Comment by evykassirer** [comment]
perfect! 

what next?

**Comment by kevinbarabash** [comment]
> what next?

I'd like to fix https://github.com/kevinbarabash/math-parser/issues/21 before we start integrating `math-parser`.  Ideally, there should be a `parse` function that can be imported instead of importing a constructor and creating an instance of `Parser`.

After that comes the fun stuff... adding `math-parser` as a submodule, switching over `parse` and `evaluate` and seeing how much stuff breaks (hopefully it's not a lot).  :crossed_fingers:  It's probably best to do this in a branch.

**Comment by kevinbarabash** [comment]
@evykassirer https://github.com/kevinbarabash/math-parser/issues/21 is complete.

**Comment by evykassirer** [comment]
sweet, now the fun stuff

do you want to take a crack at it? it'd be your first mathsteps PR!

another thing I thought of that we'll need is constructors to replace `Node.Creator`

**Comment by kevinbarabash** [comment]
I'll give it a go this weekend.  In terms of node constructors, there's a `nodes` object that `math-parser` exports which has functions for create nodes.  Unfortunately, it creates `math-ast` nodes instead of `mathjs`-style nodes.  Since the nodes are just data, we can create them right in `Node.Creator`.

**Comment by evykassirer** [comment]
from #144: 

> There are a number of test failures that are proving really tricky to fix. Fixing one test breaks others that I fixed previously. :( This is a combination of me not having a comprehensive understanding of all of the different search and transform routines and their order and trying to work at too low a level of abstraction.

> I'm going to put this PR on hold while I look at creating matcher functions as described in #142 for math-ast. I think having these functions will result in code that's easier to reason about and easier to maintain.

> I'll probably start working on the matcher inside of math-parser and then split it out into its own repo afterwards. The math-parser itself will need to be able to recognize placeholder nodes (with qualifiers). I think it makes sense to include a description of what those nodes look like in the math-ast spec.
